no-downtime-upgrade/src/main/java/org/keycloak/playground/nodowntimeupgrade/base/storage/Storage.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.playground.nodowntimeupgrade.base.storage;
+
+import org.keycloak.playground.nodowntimeupgrade.base.model.HasId;
+import java.util.Set;
+
+/**
+ *
+ * @author hmlnarik
+ */
+public interface Storage<ModelType extends HasId<Integer>> {
+    /**
+     * Creates an object in the store. Introduces new ID.
+     * @param object
+     * @return ID of the object
+     */
+    Integer create(ModelType object);
+
+    /**
+     * Returns object with the given {@code id} from the database or {@code null} if object does not exist.
+     * @param id
+     * @return
+     */
+    ModelType read(Integer id);
+
+    /**
+     * Writes object with the given {@code id} to the database.
+     * @param id
+     * @return
+     * @throws NullPointerException if object or its {@code id} is {@code null}
+     */
+    void write(Integer id, ModelType object);
+
+    /**
+     * Deletes object with the given {@code id} from the database, if exists, no-op otherwise.
+     * @param id
+     */
+    void delete(Integer id);
+
+    /**
+     * Returns set of keys of objects currently present in the store.
+     * @return
+     */
+    Set<Integer> keys();
```

Comment:
Do we really want the method like this or is it here just for the prototyping purposes? I afraid that it won't scale. Will be good to add some support for pagination and filtering etc, but I suppose you plan to do this in the later stages of the prototype?

Reply:
Just for prototyping - it is an auxiliary function. For multi-object read, there will be a specific function that would read objects depending on some criteria (see initial comment for more info on this idea)

Reply:
Thanks for clarifying this and also thanks for clarification in your other comments.